packages/network/src/logic/proxy/ProxyStreamConnectionClient.ts

```diff
@@ -0,0 +1,289 @@
+    private readonly connections: Map<StreamPartID, Map<NodeId, ProxyDirection>>
+
+    constructor(opts: ProxyStreamConnectionClientOptions) {
+        super()
+        this.trackerManager = opts.trackerManager
+        this.streamPartManager = opts.streamPartManager
+        this.nodeToNode = opts.nodeToNode
+        this.node = opts.node
+        this.nodeConnectTimeout = opts.nodeConnectTimeout
+        this.propagation = opts.propagation
+        this.definitions = new Map()
+        this.connections = new Map()
+        this.abortController = new AbortController()
+        this.nodeToNode.on(NodeToNodeEvent.PROXY_CONNECTION_RESPONSE_RECEIVED, (message, nodeId) => {
+            this.processHandshakeResponse(message, nodeId)
+        })
+    }
+
+    public async setProxies(
+        streamPartId: StreamPartID,
+        nodeIds: NodeId[],
+        direction: ProxyDirection,
+        getUserId: () => Promise<string>,
+        connectionCount?: number
+    ): Promise<void> {
+        logger.trace(`Set proxies on ${streamPartId}`)
+        this.definitions.set(streamPartId, {
+            nodeIds: new Set(nodeIds),
+            userId: await getUserId(),
+            direction,
+            connectionCount: connectionCount ?? nodeIds.length
+        })
+        await this.updateConnections(streamPartId)
+    }
+
+    private async updateConnections(streamPartId: StreamPartID): Promise<void> {
+        if (!this.definitions.has(streamPartId)) {
+            return
+        }
+        await Promise.all(this.getInvalidConnections(streamPartId).map(async (id) => {
+            logger.debug(`Closing invalid connection to ${id} on ${streamPartId}`)
+            await this.closeConnection(streamPartId, id)
+        }))
+        const connectionCountDiff =  this.definitions.get(streamPartId)!.connectionCount - this.getConnections(streamPartId).size
+        if (connectionCountDiff > 0) {
+            await this.openRandomConnections(streamPartId, connectionCountDiff)
+        } else if (connectionCountDiff < 0) {
+            await this.closeRandomConnections(streamPartId, -connectionCountDiff)
+        }
+    }
+
+    private getInvalidConnections(streamPartId: StreamPartID): string[] {
+        return Array.from(this.getConnections(streamPartId).keys()).filter((id) =>
+            !this.definitions.get(streamPartId)!.nodeIds.has(id)
+            || this.definitions.get(streamPartId)!.direction !== this.getConnections(streamPartId).get(id)
+        )
+    }
+
+    private async openRandomConnections(streamPartId: StreamPartID, connectionCount: number): Promise<void> {
+        logger.debug(`Open ${connectionCount} random connections on ${streamPartId}`)
+        const definition = this.definitions.get(streamPartId)!
+        const proxiesToAttempt = shuffle(Array.from(definition.nodeIds.keys()).filter((id) =>
+            !this.getConnections(streamPartId).has(id)
+        )).map((id) => id).splice(0, connectionCount)
```

Comment:
```suggestion
        const candidates = Array.from(definition.nodeIds.keys()).filter((id) => !this.getConnections(streamPartId).has(id)
        const proxiesToAttempt = sampleSize(candidates, connectionCount)
```

Similar also to line 217.  Uses `sampleSize` from Lodash: https://lodash.com/docs/4.17.15#sampleSize